### PR TITLE
Added mutex wait/release callback hooks for thread safety.

### PR DESCRIPTION
**Pull request by tommikas**
Also added very simple unit tests and added yotta build/test generated stuff to .mbedignore.

### CODE REVIEW

**Comment by tommikas** [comment]
@jupe @teetak01 @SeppoTakalo Check this please.

**Comment by tommikas** [comment]
Usage example pending push access so I can push and create a PR to link.

File: source/mbed_trace.c [hunk 1]
```diff
@@ -267,7 +290,10 @@ void mbed_tracef(uint8_t dlevel, const char *grp, const char *fmt, ...)
 }
 void mbed_vtracef(uint8_t dlevel, const char* grp, const char *fmt, va_list ap)
 {
+    MUTEX_WAIT(); //MUTEX_RELEASE() must be called before returning
+
     if (NULL == m_trace.line) {
+        MUTEX_RELEASE();
```
Comment (RomanSaveljev) [inline source/mbed_trace.c]: maybe a (evil) `goto` would be more maintainable here

Comment (SeppoTakalo) [inline source/mbed_trace.c]: In Nanostack, we use lots of `goto` statements. They have a place for these kind of error handling.

```
void f(...)
{
    MUTEX_WAIT();

    if (...) { // There might be multiple cases where we need to return.
        goto end;
    }
    ...rest of the function...
end:
    MUTEX_RELEASE();
}
```

In Nanostack, those case are usually `drop:` to mean that we dropped this packet, and there might be multiple reasons to do so, but only one way to drop/free it.

Comment (RomanSaveljev) [inline source/mbed_trace.c]: Thanks for clarifying my vague comment :) It was exactly my point

Comment (jupe) [inline source/mbed_trace.c]: `goto` is fine for me

**Comment by jupe** [comment]
please update readme as well

**Comment by jupe** [comment]
those helper functions needs also protection so that buffers doesnt mess when calling from threads..

**Comment by tommikas** [comment]
Good point... At first glance I thought this was enough since those helpers are always called from within the trace call, but of course the helpers actually get called first.

**Comment by tommikas** [comment]
Putting locks in the helper functions still leaves the possibility of another thread acquiring the lock in between the helper function releasing it and the actual trace function acquiring it, if the thread gets preempted.

Having the lock in place from first entering into the first called helper function until returning from the trace call itself would prevent that. One way to achieve that would be to use a counting mutex that is acquired - but not released - in helper functions, and then completely released before returning from the actual trace call. Unfortunately that means needing to have recursive mutexes available, instead of just a simple mutex.

**Comment by tommikas** [comment]
Also changed relevant things brought up in ARMmbed/mbed-client-cli#42.

File: README.md [hunk 2]
```diff
@@ -26,7 +26,8 @@ The purpose of the library is to provide a light, simple and general tracing sol
 * The trace function uses `stdout` as the default output target because it goes directly to serial port when initialized. 
 * The trace function produces traces like: `[<levl>][grp ]: msg`. This provides an easy way to detect trace prints and separate traces from normal prints (for example with _regex_).
 * This approach requires a `sprintf` implementation (`stdio.h`). The memory consumption is pretty high, but it allows an efficient way to format traces.
-* The solution is not Thread nor Interrupt safe. (Sorry, but the current implementation does not have mutexes. PRs are more than welcome.)
+* The solution is not Interrupt safe (Sorry, but the current implementation does not have internally defined mutexes. PRs are more than welcome.)
```
Comment (jupe) [inline README.md]: `implementation does not have internally defined mutexes` could be replaced with other more relevant text..

Comment (tommikas) [inline README.md]: Agreed, but I'm not quite sure what exactly to say there. Maybe just "The solution is not Interrupt safe. (PRs are more than welcome.)"?

**Comment by kjbracey-arm** [comment]
The lock thing's a bit scary, but can't see a better alternative. Does mean that an external buffer-sharing "helper" would need API to grab that.

Feels a bit like the mutex should be implicit in tmp_data_left(), or explicitly acquired by some sort of tmp_data-type call, that external helpers would get access to.

**Comment by tommikas** [comment]
Adding acquiring the lock into tmp_data_left() as it is would ensure that it gets done whenever actually needed, as long as tmp_data_left() is used correctly. The name of the macro would seem ill-fitting with that baked in though.

From the point of view of the hypothetical external helper API it'd simplify things though - a single call that locks if needed and returns number of bytes left in the buffer before starting to do anything else. Of course such a call and the current implementation internal to mbed-trace wouldn't be mutually exclusive, so it'd be easy to add whenever.

**Comment by teetak01** [comment]
+1

**Comment by jupe** [comment]
+1

**Pull request merged**